src/WinRT.Runtime/Interop/IContextCallback.cs

```diff
@@ -39,25 +65,18 @@ public static void ContextCallback(IntPtr contextCallbackPtr, delegate*<object,
             comCallData.dwDispid = 0;
             comCallData.dwReserved = 0;
 
-            CallbackData callbackData;
-            callbackData.Callback = callback;
-            callbackData.State = state;
+            CallbackData callbackData = CallbackData.GetOrCreate();
 
-            // We can just store a pointer to the callback to invoke in the context,
-            // so we don't need to allocate another closure or anything. The callback
-            // will be kept alive automatically, because 'comCallData' is address exposed.
-            // We only do this if we can use C# 11, and if we're on modern .NET, to be safe.
-            // In the callback below, we can then just retrieve the Action again to invoke it.
-            comCallData.pUserDefined = (IntPtr)(void*)&callbackData;
-            
+            comCallData.pUserDefined = GCHandle.ToIntPtr(callbackData.Handle);

```

Comment:
As discussed in Discord, you should be able to do it without GCHandle like so (you can fix up the comments, etc.):
```diff
#if NET && CsWinRT_LANG_11_FEATURES
    internal unsafe struct CallbackData
    {
+       [ThreadStatic] public static object PerThreadObject = null!;
        public delegate*<object, void> Callback;
-       public object State;
+       public object* StatePtr;
    }
#endif

#if NET && CsWinRT_LANG_11_FEATURES
    internal unsafe struct IContextCallbackVftbl
    {
#pragma warning disable CS0649 // Native layout
        private global::WinRT.Interop.IUnknownVftbl IUnknownVftbl;
        private delegate* unmanaged[Stdcall]<IntPtr, IntPtr, ComCallData*, Guid*, int, IntPtr, int> ContextCallback_4;
#pragma warning restore CS0649

        public static void ContextCallback(IntPtr contextCallbackPtr, delegate*<object, void> callback, delegate*<object, void> onFailCallback, object state)
        {
            ComCallData comCallData;
            comCallData.dwDispid = 0;
            comCallData.dwReserved = 0;

+           CallbackData.PerThreadObject = state; //you may need some sort of barrier/volatile here depending on what memory ordering the later call provides
+           fixed (object* ptr = &CallbackData.PerThreadObject)
+           {
            CallbackData callbackData;
            callbackData.Callback = callback;
+           callbackData.StatePtr = ptr;
-           callbackData.State = state;

            // We can just store a pointer to the callback to invoke in the context,
            // so we don't need to allocate another closure or anything. The callback
            // will be kept alive automatically, because 'comCallData' is address exposed.
            // We only do this if we can use C# 11, and if we're on modern .NET, to be safe.
            // In the callback below, we can then just retrieve the Action again to invoke it.
            comCallData.pUserDefined = (IntPtr)(void*)&callbackData;
            
            [UnmanagedCallersOnly]
            static int InvokeCallback(ComCallData* comCallData)
            {
                try
                {
                    CallbackData* callbackData = (CallbackData*)comCallData->pUserDefined;

-                   callbackData->Callback(callbackData->State);
+                   callbackData->Callback(*callbackData->StatePtr);

                    return 0; // S_OK
                }
                catch (Exception e)
                {
                    return e.HResult;
                }
            }

            Guid iid = IID.IID_ICallbackWithNoReentrancyToApplicationSTA;

            int hresult = (*(IContextCallbackVftbl**)contextCallbackPtr)->ContextCallback_4(
                contextCallbackPtr,
                (IntPtr)(delegate* unmanaged<ComCallData*, int>)&InvokeCallback,
                &comCallData,
                &iid,
                /* iMethod */ 5,
                IntPtr.Zero);
+           }
+           CallbackData.PerThreadObject = null!; //you may need some sort of barrier/volatile here depending on what memory ordering the above call provides

            if (hresult < 0)
            {
                if (onFailCallback is not null)
                {
                    onFailCallback(state);
                }
            }
        }
    }
```

Reply:
Mmh I mean I like the idea of not needing a `GCHandle` and the finalizer and the `GC.KeepAlive` call, but this feels even trickier and I'd like someone else to double check that it would be correct and not relying on some other UB of some kind. @jkotas if you have a minute would you mind taking a look and sharing your thoughts? Thank you! 😅

Reply:
I do not see a problem with this suggestion.